caddyhttp/httpserver/replacer.go

```diff
@@ -309,6 +309,12 @@ func (r *replacer) getSubstitution(key string) string {
 	case "{dir}":
 		dir, _ := path.Split(r.request.URL.Path)
 		return dir
+	case "{user}":
+		user, _ := r.request.Context().Value(RemoteUserCtxKey).(string)
+		if user == "" {
+			user = "-"
```

Comment:
@jamesboswell `"-"` **is** the default empty value.

See the third argument to NewReplacer:

https://github.com/mholt/caddy/blob/6cabc9bfe3dc5f06c49aa8fb58ac19a09a4f42e9/caddyhttp/httpserver/replacer.go#L90

And the log always passes in `"-"`:

https://github.com/mholt/caddy/blob/6cabc9bfe3dc5f06c49aa8fb58ac19a09a4f42e9/caddyhttp/log/log.go#L34

Consequently, the line highlighted in this PR will break that argument if some middleware ever wants to use something other than "-" (which some do, some just want an empty string because they aren't always using this for logging).